File: internal/verdict/verdict.go
// Package verdict scores enriched survivors with the Anthropic API using the
// current taste block. Results are persisted to verdicts. A scored company is
// sticky: a default run skips any company that already has a verdict (criteria
// or playbook edits do not re-score it). Re-scoring is always explicit — a
// targeted per-company run, or a --force run that re-scores everything.
package verdict

import (
	"context"
	"encoding/json"
	"fmt"
	"os"
	"regexp"
	"strings"
	"sync"
	"time"

	"github.com/slaguardia/scout/internal/anthropic"
	"github.com/slaguardia/scout/internal/filter"
	"github.com/slaguardia/scout/internal/store"
	"github.com/slaguardia/scout/internal/taste"
)

// Scorer is the verdict driver.
type Scorer struct {
	DB     *store.DB
	Taste  *taste.Block
	Filter *filter.Taste
	Client *anthropic.Client
	Model  string
	Force  bool // re-score every eligible company, replacing existing verdicts

	// OnlyBlanks limits the run to companies with no verdict row at all — the
	// "just the new arrivals" pass. This is also what a default (non-force) run
	// now does, since scored companies are sticky; the flag stays for explicit
	// callers and to share the run-dialog knob with enrich. Takes precedence
	// over Force.
	OnlyBlanks bool

	// CompanyIDs limits the run to exactly these companies and always
	// re-scores them — a targeted run is an explicit ask, so up-to-date and
	// even manual verdicts are overwritten. Overrides Force and OnlyBlanks.
	// The explicit ask also bypasses the static taste filter: a company you
	// point at by id is scored even if it wouldn't survive the bulk pre-filter
	// (e.g. an excluded-vertical substring false-positive). Only enrichment
	// eligibility still applies — a company lacking an 'ok' row is reported,
	// not scored, since there's no fetched text to reason over.
	CompanyIDs []string

	// Playbook is the agent's operating manual (how to decide) — distinct from
	// Taste (what the user wants). Empty means fall back to the built-in rubric.
	// It's recorded in Taste.Version (the criteria the verdict was scored under,
	// shown in the trail), but editing it no longer re-scores existing verdicts.
	Playbook string

	// RunID, when set, tags every decision-trail row with the UI run uuid so a
	// company's timeline can be grouped by run. Empty for CLI runs.
	RunID string

	Workers int

	// Progress, if set, receives one line per scored company (both passes).
	// Called from worker goroutines — must be safe for concurrent use.
	Progress func(string)
}

func (s *Scorer) emit(line string) {
	if s.Progress != nil {
		s.Progress(line)
	}
}

// Result is the run summary.
type Result struct {
	Considered          int
	Scored              int
	Skipped             int
	Failed              int
	ByVerdict           map[string]int
	CacheCreationTokens int // sum of cache_creation_input_tokens across all calls
	CacheReadTokens     int // sum of cache_read_input_tokens (the saving)
}

// Run scores every survivor (per filter rules) that has enrichment and lacks
// an up-to-date verdict.
func (s *Scorer) Run(ctx context.Context) (*Result, error) {
	if s.Workers <= 0 {
		s.Workers = 4
	}
	if s.Model == "" {
		s.Model = anthropic.DefaultModel
	}

	cands, err := s.candidates()
	if err != nil {
		return nil, err
	}

	res := &Result{Considered: len(cands), ByVerdict: map[string]int{}}
	if len(cands) == 0 {
		return res, nil
	}

	// Header up front so the parallelism is legible: the per-company lines below
	// stream one at a time as each finishes, but the work runs Workers-at-once.
	workers := s.Workers
	if workers > len(cands) {
		workers = len(cands)
	}
	s.emit(fmt.Sprintf("scoring %d companies · %d workers in parallel", len(cands), workers))

	jobs := make(chan store.VerdictCandidate)
	var wg sync.WaitGroup
	var mu sync.Mutex

	for i := 0; i < s.Workers; i++ {
		wg.Add(1)
		go func() {
			defer wg.Done()
			for c := range jobs {
				s.emit(fmt.Sprintf("· %s…", c.Name)) // picked up — shows the in-flight burst
				v, cacheCreate, cacheRead, err := s.scoreOne(ctx, c)
				mu.Lock()
				res.CacheCreationTokens += cacheCreate
				res.CacheReadTokens += cacheRead
				if err != nil {
					res.Failed++
					mu.Unlock()
					fmt.Printf("verdict %s (%s) error: %v\n", c.CompanyID, c.Name, err)
					s.emit(fmt.Sprintf("%s — error: %v", c.Name, err))
					continue
				}
				if v == nil {
					res.Skipped++
					mu.Unlock()
					s.emit(fmt.Sprintf("%s — skipped (up to date)", c.Name))
					continue
				}
				res.Scored++
				res.ByVerdict[v.Verdict]++
				mu.Unlock()
				s.emit(fmt.Sprintf("%s → %s — %s", c.Name, v.Verdict, v.Reason))
			}
		}()
	}

	for _, c := range cands {
		select {
		case <-ctx.Done():
			close(jobs)
			wg.Wait()
			return res, ctx.Err()
		case jobs <- c:
		}
	}
	close(jobs)
	wg.Wait()
	return res, nil
}

// candidates returns the companies to score, each paired with its 'ok'
// enrichment summary. A bulk run scores everything that survives the static
// taste filter; a targeted run scores exactly the requested companies and
// bypasses the filter entirely — an explicit re-score is a deliberate ask, and
// taste.toml is only a cheap bulk pre-filter to save LLM cost, not a veto on a
// company you've pointed at by id. Enrichment eligibility ('ok' row) still
// applies in both cases: with no fetched text there's nothing to reason over.
func (s *Scorer) candidates() ([]store.VerdictCandidate, error) {
	wanted := make(map[string]bool, len(s.CompanyIDs))
	for _, id := range s.CompanyIDs {
		wanted[id] = true
	}

	var ids []string
	var byID map[string]filter.Survivor

	if len(wanted) > 0 {
		// Targeted: load the requested companies straight from the table,
		// skipping the static filter. Say why anything asked for is missing — a
		// silent zero-company run reads as a bug.
		svs, err := s.requestedCompanies(s.CompanyIDs)
		if err != nil {
			return nil, err
		}
		byID = make(map[string]filter.Survivor, len(svs))
		for _, sv := range svs {
			ids = append(ids, sv.ID)
			byID[sv.ID] = sv
		}
		if len(ids) < len(wanted) {
			s.emit(fmt.Sprintf("targeted: %d of %d requested companies exist", len(ids), len(wanted)))
		}
	} else {
		fres, err := s.Filter.Apply(s.DB)
		if err != nil {
			return nil, err
		}
		byID = make(map[string]filter.Survivor, len(fres.Survivors))
		for _, sv := range fres.Survivors {
			ids = append(ids, sv.ID)
			byID[sv.ID] = sv
		}
	}
	if len(ids) == 0 {
		return nil, nil
	}

	// Pull enrichment summaries for those IDs.
	q, args := buildInQuery(`
SELECT company_id, COALESCE(website_summary, '')
FROM enrichment
WHERE fetch_status = 'ok' AND company_id IN `, ids)
	rows, err := s.DB.Query(q, args...)
	if err != nil {
		return nil, err
	}
	defer rows.Close()

	var out []store.VerdictCandidate
	for rows.Next() {
		var id string
		var summary string
		if err := rows.Scan(&id, &summary); err != nil {
			return nil, err
		}
		sv := byID[id]
		out = append(out, store.VerdictCandidate{
			CompanyID:      id,
			Name:           sv.Name,
			Domain:         sv.Domain,
			Location:       sv.Location,
			Vertical:       sv.Vertical,
			Headcount:      sv.Headcount,
			Stage:          sv.Stage,
			WebsiteSummary: summary,
		})
	}
	if len(wanted) > 0 && len(out) < len(ids) {
		s.emit(fmt.Sprintf("targeted: %d of %d requested companies have an ok enrichment row", len(out), len(ids)))
	}
	return out, rows.Err()
}

// requestedCompanies loads the given companies as filter.Survivor projections,
// bypassing the static taste filter — used only by targeted runs, where the
// explicit ask overrides the bulk pre-filter. The SELECT mirrors
// filter.Taste.Apply so the user prompt is built from identical fields. IDs
// with no matching row are simply absent (the caller reports the shortfall).
func (s *Scorer) requestedCompanies(idList []string) ([]filter.Survivor, error) {
	q, args := buildInQuery(`
SELECT id, name, COALESCE(domain,''), COALESCE(location,''), COALESCE(vertical,''),
       COALESCE(headcount, 0), COALESCE(funding_stage,'')
FROM companies WHERE id IN `, idList)
	rows, err := s.DB.Query(q, args...)
	if err != nil {
		return nil, err
	}
	defer rows.Close()

	var out []filter.Survivor
	for rows.Next() {
		var sv filter.Survivor
		if err := rows.Scan(&sv.ID, &sv.Name, &sv.Domain, &sv.Location, &sv.Vertical, &sv.Headcount, &sv.Stage); err != nil {
			return nil, err
		}
		out = append(out, sv)
	}
	return out, rows.Err()
}

func buildInQuery(prefix string, ids []string) (string, []any) {
	if len(ids) == 0 {
		return prefix + "()", nil
	}
	ph := make([]string, len(ids))
	args := make([]any, len(ids))
	for i, id := range ids {
		ph[i] = "?"
		args[i] = id
	}
	return prefix + "(" + strings.Join(ph, ",") + ")", args
}

// scoreOne returns the upserted verdict (nil if skipped), plus the
// cache_creation and cache_read input-token counts from the response so
// Run() can aggregate them.
func (s *Scorer) scoreOne(ctx context.Context, c store.VerdictCandidate) (*store.Verdict, int, int, error) {
	// A targeted run always re-scores — the user pointed at this company on
	// purpose, so even a sticky manual verdict is fair game.
	if len(s.CompanyIDs) == 0 && (s.OnlyBlanks || !s.Force) {
		existing, err := s.DB.GetVerdict(c.CompanyID)
		if err != nil {
			return nil, 0, 0, err
		}
		if existing != nil {
			// Any already-scored company is left untouched on a default or
			// blanks-only run. Verdicts don't go stale on a criteria change — they
			// persist until an explicit re-score (--force, or a targeted run). This
			// keeps a brief/playbook edit from silently churning every verdict.
			return nil, 0, 0, nil
		}
	}

	system := buildSystemPrompt(s.Playbook, s.Taste.Text)
	user := buildUserPrompt(c)

	// Bound per-call latency separately from the global ctx.
	callCtx, cancel := context.WithTimeout(ctx, 45*time.Second)
	defer cancel()

	resp, err := s.Client.Send(callCtx, anthropic.Request{
		Model:     s.Model,
		System:    system,
		MaxTokens: 256,
		Messages:  []anthropic.Message{{Role: "user", Content: user}},
		Cached:    true, // taste + rubric are identical across all calls in a run
	})
	if err != nil {
		return nil, 0, 0, err
	}

	verdict, reason, err := parseVerdict(resp.Text())
	if err != nil {
		return nil, resp.Usage.CacheCreationInputTokens, resp.Usage.CacheReadInputTokens,
			fmt.Errorf("parse: %w (raw=%q)", err, truncate(resp.Text(), 200))
	}

	v := store.Verdict{
		CompanyID:    c.CompanyID,
		Verdict:      verdict,
		Reason:       reason,
		TasteVersion: s.Taste.Version,
		Model:        s.Model,
	}
	if err := s.DB.UpsertVerdict(v); err != nil {
		return nil, resp.Usage.CacheCreationInputTokens, resp.Usage.CacheReadInputTokens, err
	}
	s.writeTrace(c, s.Model, verdict, reason)
	return &v, resp.Usage.CacheCreationInputTokens, resp.Usage.CacheReadInputTokens, nil
}

// hardContract is the one invariant the parser depends on. It is never
// editable from the playbook — a broken output contract breaks parsing.
const hardContract = `You are Scout's verdict engine. Given a company, decide if it's worth the user's time to investigate further as a job opportunity. Reply ONLY with valid JSON, no preamble, no markdown fences. The JSON must have exactly two fields:
  {"verdict": "yes"|"maybe"|"no", "reason": "one-line, specific"}`

// builtinRubric is the fallback "how to decide" guidance used only when no
// playbook is supplied. The shipped default playbook supersedes this.
const builtinRubric = `Verdict rubric:
  - "yes":   high-confidence fit. Worth the user actively investigating.
  - "maybe": adjacent or uncertain. Worth a skim, not a deep dive.
  - "no":    poor fit or hard exclusion.

The reason must be specific — name the vertical, stage, or trait that drove the call. Don't say "matches taste" or "good fit"; say "AI infra for ML teams, Series B" or "crypto wallet (excluded)".`

// hardGateRubric tells the LLM how to read the criteria brief. The brief is a
// distilled, prose company-fit summary (from the distiller, or taste.md
// offline) — there are no [requires]/[excludes] tags to key on; the stance is
// in the words. The brief states acceptable alternatives explicitly ("any one
// of: X, Y, Z"), so the OR-set logic lives in the brief's prose, and this rubric
// only has to say how to weigh dealbreakers vs preferences vs context.
const hardGateRubric = `The criteria below are a distilled company-fit brief in the user's own terms. Read it and apply it like this:
• Anything stated as a hard dealbreaker or exclusion is a gate: if the company hits it, the verdict is "no" (red). Name the dealbreaker in the reason.
• Anything stated as a hard requirement is a gate that must hold on its own. Where the brief lists acceptable alternatives ("any one of: X, Y, Z"), matching ONE satisfies it — not matching the others is expected and is NOT a strike.
• Strong preferences are weights, not gates: a miss leans "maybe" (yellow), never an automatic "no".
• Context is background for judgment, not a rule to gate on.

`

// buildSystemPrompt assembles three layers: the hard JSON contract (fixed),
// the playbook / how-to-decide (operator-editable, falls back to the builtin
// rubric), then the criteria / what-the-user-wants block.
func buildSystemPrompt(playbook, criteria string) string {
	var b strings.Builder
	b.WriteString(hardContract)

	b.WriteString("\n\n--- PLAYBOOK (how to decide) ---\n")
	if pb := strings.TrimSpace(playbook); pb != "" {
		b.WriteString(pb)
	} else {
		b.WriteString(builtinRubric)
	}

	b.WriteString("\n\n--- CRITERIA (what the user wants) ---\n")
	b.WriteString(hardGateRubric)
	b.WriteString(strings.TrimSpace(criteria))
	return b.String()
}

func buildUserPrompt(c store.VerdictCandidate) string {
	var b strings.Builder
	fmt.Fprintf(&b, "Company: %s\n", c.Name)
	if c.Domain != "" {
		fmt.Fprintf(&b, "Domain: %s\n", c.Domain)
	}
	if c.Vertical != "" {
		fmt.Fprintf(&b, "Vertical: %s\n", c.Vertical)
	}
	if c.Location != "" {
		fmt.Fprintf(&b, "Location: %s\n", c.Location)
	}
	if c.Headcount > 0 {
		fmt.Fprintf(&b, "Headcount: %d\n", c.Headcount)
	}
	if c.Stage != "" {
		fmt.Fprintf(&b, "Funding stage: %s\n", c.Stage)
	}
	if c.WebsiteSummary != "" {
		fmt.Fprintf(&b, "\nWebsite text (truncated):\n%s\n", c.WebsiteSummary)
	}
	b.WriteString("\nReturn the JSON verdict now.")
	return b.String()
}

// writeTrace appends one decision-trail row for a completed scoring pass: which
// criteria/version/model drove this verdict. Best-effort — a failure is logged
// but never fails the verdict (the trail is a debugging aid, not the result).
func (s *Scorer) writeTrace(c store.VerdictCandidate, model, verdict, reason string) {
	t := store.VerdictTrace{
		CompanyID:      c.CompanyID,
		RunID:          s.RunID,
		Model:          model,
		TasteVersion:   s.Taste.Version,
		CriteriaSource: s.Taste.Source,
		Verdict:        verdict,
		Reason:         reason,
	}
	if err := s.DB.InsertVerdictTrace(t); err != nil {
		fmt.Fprintf(os.Stderr, "verdict trace %s (%s): %v\n", c.CompanyID, c.Name, err)
	}
}

// Verdict parsing: tolerant of surrounding noise, fenced code blocks, etc.

var reJSON = regexp.MustCompile(`(?s)\{[^{}]*\}`)

func parseVerdict(s string) (verdict, reason string, err error) {
	s = strings.TrimSpace(s)
	candidates := []string{s}
	if m := reJSON.FindString(s); m != "" {
		candidates = append([]string{m}, candidates...)
	}
	for _, c := range candidates {
		var v struct {
			Verdict string `json:"verdict"`
			Reason  string `json:"reason"`
		}
		if err := json.Unmarshal([]byte(c), &v); err == nil {
			vv := strings.ToLower(strings.TrimSpace(v.Verdict))
			switch vv {
			case "yes", "maybe", "no":
				return vv, strings.TrimSpace(v.Reason), nil
			}
		}
	}
	return "", "", fmt.Errorf("no valid verdict JSON")
}

func truncate(s string, n int) string {
	if len(s) <= n {
		return s
	}
	return s[:n] + "…"
}

File: scout/verdict/verdict.py
"""Score enriched survivors with the Anthropic API over the current criteria
brief.

Results are persisted to verdicts. A scored company is sticky: a default run skips
any company that already has a verdict (criteria or playbook edits do not re-score
it). Re-scoring is always explicit — a targeted per-company run, or a force run.

Concurrency note: scoring runs sequentially because the single shared sqlite3
connection is not thread-safe. Result accounting, the verdict + trace writes, and
the progress emissions are unaffected — only wall-clock parallelism.
"""

from __future__ import annotations

import json
import re
import sys
from collections.abc import Callable
from dataclasses import dataclass, field

from scout import anthropic, taste
from scout.filter import Survivor, Taste
from scout.store.trace import VerdictTrace, insert_verdict_trace
from scout.store.verdicts import Verdict, VerdictCandidate, get_verdict, upsert_verdict


@dataclass
class Result:
    """The run summary."""

    considered: int = 0
    scored: int = 0
    skipped: int = 0
    failed: int = 0
    by_verdict: dict[str, int] = field(default_factory=dict)
    cache_creation_tokens: int = 0  # sum of cache_creation_input_tokens across all calls
    cache_read_tokens: int = 0  # sum of cache_read_input_tokens (the saving)


class Scorer:
    """The verdict driver."""

    def __init__(
        self,
        con=None,
        taste: taste.Block | None = None,
        filter: Taste | None = None,
        client: anthropic.Client | None = None,
        model: str = "",
        force: bool = False,
        only_blanks: bool = False,
        company_ids: list[str] | None = None,
        playbook: str = "",
        run_id: str = "",
        workers: int = 0,
        progress: Callable[[str], None] | None = None,
    ):
        self.con = con
        self.taste = taste
        self.filter = filter
        self.client = client
        self.model = model
        self.force = force  # re-score every eligible company, replacing existing verdicts
        # only_blanks limits the run to companies with no verdict row at all. Takes
        # precedence over force.
        self.only_blanks = only_blanks
        # company_ids limits the run to exactly these companies and always
        # re-scores them, bypassing the static taste filter. Overrides force/only_blanks.
        self.company_ids = company_ids
        # playbook is the agent's operating manual (how to decide). Empty falls
        # back to the built-in rubric.
        self.playbook = playbook
        # run_id tags every decision-trail row with the UI run uuid. Empty for CLI.
        self.run_id = run_id
        self.workers = workers
        self.progress = progress

    def emit(self, line: str) -> None:
        if self.progress is not None:
            self.progress(line)

    def run(self) -> Result:
        """Score every survivor (per filter rules) that has enrichment and lacks
        an up-to-date verdict."""
        if self.workers <= 0:
            self.workers = 4
        if self.model == "":
            self.model = anthropic.DEFAULT_MODEL

        cands = self.candidates()
        res = Result(considered=len(cands))
        if not cands:
            return res

        # Header up front so the parallelism is legible.
        workers = min(self.workers, len(cands))
        self.emit(f"scoring {len(cands)} companies · {workers} workers in parallel")

        for c in cands:
            self.emit(f"· {c.name}…")  # picked up
            v, cache_create, cache_read, err = self.score_one(c)
            res.cache_creation_tokens += cache_create
            res.cache_read_tokens += cache_read
            if err is not None:
                res.failed += 1
                print(f"verdict {c.company_id} ({c.name}) error: {err}")
                self.emit(f"{c.name} — error: {err}")
                continue
            if v is None:
                res.skipped += 1
                self.emit(f"{c.name} — skipped (up to date)")
                continue
            res.scored += 1
            res.by_verdict[v.verdict] = res.by_verdict.get(v.verdict, 0) + 1
            self.emit(f"{c.name} → {v.verdict} — {v.reason}")
        return res

    def candidates(self) -> list[VerdictCandidate]:
        """The companies to score, each paired with its 'ok' enrichment summary. A
        bulk run scores everything that survives the static taste filter; a
        targeted run scores exactly the requested companies and bypasses the filter
        entirely. Enrichment eligibility ('ok' row) still applies in both cases."""
        wanted = set(self.company_ids or [])

        if wanted:
            # Targeted: load the requested companies straight from the table,
            # skipping the static filter.
            svs = self.requested_companies(self.company_ids)
            by_id = {sv.id: sv for sv in svs}
            ids = [sv.id for sv in svs]
            if len(ids) < len(wanted):
                self.emit(f"targeted: {len(ids)} of {len(wanted)} requested companies exist")
        else:
            fres = self.filter.apply(self.con)
            by_id = {sv.id: sv for sv in fres.survivors}
            ids = [sv.id for sv in fres.survivors]
        if not ids:
            return []

        # Pull enrichment summaries for those IDs.
        q, args = build_in_query(
            """
SELECT company_id, COALESCE(website_summary, '')
FROM enrichment
WHERE fetch_status = 'ok' AND company_id IN """,
            ids,
        )
        rows = self.con.execute(q, args).fetchall()

        out: list[VerdictCandidate] = []
        for r in rows:
            sv = by_id[r[0]]
            out.append(
                VerdictCandidate(
                    company_id=r[0],
                    name=sv.name,
                    domain=sv.domain,
                    location=sv.location,
                    vertical=sv.vertical,
                    headcount=sv.headcount,
                    stage=sv.stage,
                    website_summary=r[1],
                )
            )
        if wanted and len(out) < len(ids):
            self.emit(
                f"targeted: {len(out)} of {len(ids)} requested companies have an ok enrichment row"
            )
        return out

    def requested_companies(self, id_list: list[str]) -> list[Survivor]:
        """Load the given companies as filter.Survivor projections, bypassing the
        static taste filter. IDs with no matching row are simply absent."""
        q, args = build_in_query(
            """
SELECT id, name, COALESCE(domain,''), COALESCE(location,''), COALESCE(vertical,''),
       COALESCE(headcount, 0), COALESCE(funding_stage,'')
FROM companies WHERE id IN """,
            id_list,
        )
        rows = self.con.execute(q, args).fetchall()
        return [
            Survivor(
                id=r[0],
                name=r[1],
                domain=r[2],
                location=r[3],
                vertical=r[4],
                headcount=r[5],
                stage=r[6],
            )
            for r in rows
        ]

    def score_one(self, c: VerdictCandidate) -> tuple[Verdict | None, int, int, Exception | None]:
        """Score one company. Returns (verdict|None, cache_creation, cache_read,
        err) — None verdict + None err means skipped (already up to date). The
        cache token counts are returned even on a parse/write error so run() can
        still aggregate them."""
        # A targeted run always re-scores — the user pointed at this company on
        # purpose, so even a sticky manual verdict is fair game.
        if not self.company_ids and (self.only_blanks or not self.force):
            existing = get_verdict(self.con, c.company_id)
            if existing is not None:
                # Any already-scored company is left untouched on a default or
                # blanks-only run.
                return None, 0, 0, None

        system = build_system_prompt(self.playbook, self.taste.text)
        user = build_user_prompt(c)

        try:
            resp = self.client.send(
                anthropic.Request(
                    model=self.model,
                    system=system,
                    max_tokens=256,
                    messages=[anthropic.Message("user", user)],
                    cached=True,  # taste + rubric are identical across all calls in a run
                    timeout=45.0,
                )
            )
        except Exception as e:  # noqa: BLE001 - surface as the run's per-company failure
            return None, 0, 0, e

        cc = resp.usage.cache_creation_input_tokens
        cr = resp.usage.cache_read_input_tokens
        try:
            verdict, reason = parse_verdict(resp.text())
        except ValueError as e:
            return None, cc, cr, ValueError(f"parse: {e} (raw={truncate(resp.text(), 200)!r})")

        v = Verdict(
            company_id=c.company_id,
            verdict=verdict,
            reason=reason,
            taste_version=self.taste.version,
            model=self.model,
        )
        try:
            upsert_verdict(self.con, v)
        except Exception as e:  # noqa: BLE001
            return None, cc, cr, e
        self.write_trace(c, self.model, verdict, reason)
        return v, cc, cr, None

    def write_trace(self, c: VerdictCandidate, model: str, verdict: str, reason: str) -> None:
        """Append one decision-trail row. Best-effort — a failure is logged but
        never fails the verdict."""
        t = VerdictTrace(
            company_id=c.company_id,
            run_id=self.run_id,
            model=model,
            taste_version=self.taste.version,
            criteria_source=self.taste.source,
            verdict=verdict,
            reason=reason,
        )
        try:
            insert_verdict_trace(self.con, t)
        except Exception as e:  # noqa: BLE001
            print(f"verdict trace {c.company_id} ({c.name}): {e}", file=sys.stderr)


def build_in_query(prefix: str, ids: list[str]) -> tuple[str, list]:
    if not ids:
        return prefix + "()", []
    ph = ",".join("?" for _ in ids)
    return prefix + "(" + ph + ")", list(ids)


# hard_contract is the one invariant the parser depends on. It is never editable
# from the playbook — a broken output contract breaks parsing.
HARD_CONTRACT = """You are Scout's verdict engine. Given a company, decide if it's worth the user's time to investigate further as a job opportunity. Reply ONLY with valid JSON, no preamble, no markdown fences. The JSON must have exactly two fields:
  {"verdict": "yes"|"maybe"|"no", "reason": "one-line, specific"}"""

# builtin_rubric is the fallback "how to decide" guidance used only when no
# playbook is supplied. The shipped default playbook supersedes this.
BUILTIN_RUBRIC = """Verdict rubric:
  - "yes":   high-confidence fit. Worth the user actively investigating.
  - "maybe": adjacent or uncertain. Worth a skim, not a deep dive.
  - "no":    poor fit or hard exclusion.

The reason must be specific — name the vertical, stage, or trait that drove the call. Don't say "matches taste" or "good fit"; say "AI infra for ML teams, Series B" or "crypto wallet (excluded)"."""

# hard_gate_rubric tells the LLM how to read the criteria brief.
HARD_GATE_RUBRIC = """The criteria below are a distilled company-fit brief in the user's own terms. Read it and apply it like this:
• Anything stated as a hard dealbreaker or exclusion is a gate: if the company hits it, the verdict is "no" (red). Name the dealbreaker in the reason.
• Anything stated as a hard requirement is a gate that must hold on its own. Where the brief lists acceptable alternatives ("any one of: X, Y, Z"), matching ONE satisfies it — not matching the others is expected and is NOT a strike.
• Strong preferences are weights, not gates: a miss leans "maybe" (yellow), never an automatic "no".
• Context is background for judgment, not a rule to gate on.

"""


def build_system_prompt(playbook: str, criteria: str) -> str:
    """Assemble three layers: the hard JSON contract (fixed), the playbook /
    how-to-decide (operator-editable, falls back to the builtin rubric), then the
    criteria / what-the-user-wants block."""
    b = HARD_CONTRACT
    b += "\n\n--- PLAYBOOK (how to decide) ---\n"
    pb = (playbook or "").strip()
    b += pb if pb != "" else BUILTIN_RUBRIC
    b += "\n\n--- CRITERIA (what the user wants) ---\n"
    b += HARD_GATE_RUBRIC
    b += (criteria or "").strip()
    return b


def build_user_prompt(c: VerdictCandidate) -> str:
    parts = [f"Company: {c.name}\n"]
    if c.domain != "":
        parts.append(f"Domain: {c.domain}\n")
    if c.vertical != "":
        parts.append(f"Vertical: {c.vertical}\n")
    if c.location != "":
        parts.append(f"Location: {c.location}\n")
    if c.headcount > 0:
        parts.append(f"Headcount: {c.headcount}\n")
    if c.stage != "":
        parts.append(f"Funding stage: {c.stage}\n")
    else:
        # Make the blank explicit. Omitting the field lets a weak model fill the
        # void — e.g. inferring "Series B" from a valuation in the website text.
        parts.append(
            "Funding stage: unknown (not in the data — do NOT infer a round from a valuation, raise, or headcount)\n"
        )
    if c.website_summary != "":
        parts.append(f"\nWebsite text (truncated):\n{c.website_summary}\n")
    parts.append("\nReturn the JSON verdict now.")
    return "".join(parts)


# Verdict parsing: tolerant of surrounding noise, fenced code blocks, etc.
_RE_JSON = re.compile(r"\{[^{}]*\}", re.S)


def parse_verdict(s: str) -> tuple[str, str]:
    """Parse the verdict JSON. Returns (verdict, reason); raises ValueError when no
    valid verdict JSON is present."""
    s = s.strip()
    candidates = [s]
    m = _RE_JSON.search(s)
    if m:
        candidates.insert(0, m.group(0))
    for c in candidates:
        try:
            v = json.loads(c)
        except (ValueError, json.JSONDecodeError):
            continue
        if not isinstance(v, dict):
            continue
        vv = str(v.get("verdict", "")).strip().lower()
        if vv in ("yes", "maybe", "no"):
            return vv, str(v.get("reason", "")).strip()
    raise ValueError("no valid verdict JSON")


def truncate(s: str, n: int) -> str:
    if len(s) <= n:
        return s
    return s[:n] + "…"
